File: maxanagrams.clj
; find the combinations of letters which have the most anagrams
; word list is from 
; https://raw.githubusercontent.com/dolph/dictionary/master/enable1.txt  
(ns anagrams-sort
  (:require [clojure.string :as cs]) )

(defn main []  

(def words
  (let [ home     (System/getenv "HOME")
         filename (str home "/Downloads/enable1.txt")
         words    (->> filename slurp cs/split-lines)
        ] words )
)

;(defn negate [x] (- 0 x))

(def sortwords
  (->> (reduce (fn [acc word]
               (update acc (apply str (sort word)) conj word))
               {}
               words)
       (sort-by #(-> % second count)))
  )

(take 10 (reverse sortwords))
              
)

(time (main))

;; reduce line is equivalent to:
;; (reduce (fn [accumulator word] (update accumulator (sort word) conj word)) {})
;or just (Justin's solution)
(comment 
  (->> "/home/aaron/Downloads/enable1.txt" slurp
       clojure.string/split-lines
       (reduce #(update %1 (sort %2) conj %2) {}) 
       (sort-by (comp count second))
       last last)
)

File: maxanagrams.hs
-- find the combinations of letters which have the most anagrams
import qualified Data.Map as M
import Data.List(words,sortBy,sort,groupBy,map,intercalate)
import Text.Printf(printf)
import System.Environment(getEnv)

type Anagrams = (Int,String,[String])

printer :: Anagrams -> IO ()
printer (len,sortword,words) = putStrLn $
  printf "%2d %-12s %s" len sortword (intercalate "," words)

sortRevLength :: (Int,a,b) -> (Int,a,b) -> Ordering
sortRevLength (l1,_,_) (l2,_,_) = compare l2 l1

topAnasSort :: [String] -> [Anagrams]
topAnasSort wrds = 
  map ( \(l,x,y)->(l, x, sort y) ) $
  sortBy sortRevLength $
  map (\l -> (length l, fst (head l), map snd l)) $
  groupBy (\x y -> fst x == fst y) $
  sortBy (\(x,_) (y,_) -> compare x y) $
  map (\x -> (sort x, x)) $ wrds

topAnasMap :: [String] -> [Anagrams]
topAnasMap wrds = 
  map (\(l,x,y)->(l,x,reverse y)) $
  sortBy sortRevLength $
  map (\l -> ((length$snd l),(fst l),(snd l)) ) $
  M.toList $
  M.fromListWith (++) (map (\w -> (sort w,[w])) wrds)

-- word list is from 
-- https://raw.githubusercontent.com/dolph/dictionary/master/enable1.txt  
main :: IO () 
main = do
  home <- getEnv "HOME"
  let filename = home ++ "/Downloads/enable1.txt"
  txt <- readFile filename
  let wrds = words txt
  printer (1, "letters", ["anagrams_sort"])
  mapM_ printer $ take 10 $ topAnasSort wrds
  putStrLn ""
  printer (2, "letters", ["anagrams_map"])
  mapM_ printer $ take 10 $ topAnasMap wrds
